lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java

```diff
@@ -1,88 +1,190 @@
 package org.enso.syntax2;
 
 import java.io.File;
+import java.lang.ref.Reference;
+import java.lang.ref.SoftReference;
 import java.net.URISyntaxException;
 import java.nio.BufferUnderflowException;
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
 import java.nio.charset.StandardCharsets;
+import java.util.UUID;
 import org.slf4j.LoggerFactory;
 
-public final class Parser implements AutoCloseable {
-  private static void initializeLibraries() {
-    try {
-      System.loadLibrary("enso_parser");
-      return;
-    } catch (LinkageError err) {
-      // try harder to find the library
-    }
-    String os = System.getProperty("os.name");
-    String name;
-    if (os.startsWith("Mac")) {
-      name = "libenso_parser.dylib";
-    } else if (os.startsWith("Windows")) {
-      name = "enso_parser.dll";
-    } else {
-      name = "libenso_parser.so";
+public final class Parser {
+  private Parser() {}
+
+  public static String getWarningMessage(Warning warning) {
+    return getWarningTemplate(warning.getId());
+  }
+
+  public static long isIdentOrOperator(CharSequence input) {
+    return getWorker().isIdentOrOperator(input);
+  }
+
+  public static ByteBuffer parseInputLazy(CharSequence input) {
+    return getWorker().parseInputLazy(input);
+  }
+
+  public static Tree parse(CharSequence input) {
+    return getWorker().parse(input);
+  }
+
+  public static UUID getUuid(long metadata, long nodeOffset, long nodeLength) {
+    long high = getUuidHigh(metadata, nodeOffset, nodeLength);
+    long low = getUuidLow(metadata, nodeOffset, nodeLength);
+    if (high == 0 && low == 0) {
+      // The native interface uses the Nil UUID value as a marker to indicate that no UUID was
+      // attached.
+      // The Nil UUID will never collide with a real UUID generated by any scheme.
+      return null;
     }
+    return new UUID(high, low);
+  }
+
+  /* Worker-thread state */
 
-    var whereAmI = Parser.class.getProtectionDomain().getCodeSource().getLocation();
-    File root;
-    try {
-      root = new File(whereAmI.toURI()).getParentFile();
-    } catch (URISyntaxException ex) {
-      root = new File(".").getAbsoluteFile();
+  private static final FinalizationManager finalizationManager = new FinalizationManager();
+  private static final Thread finalizationThread = new Thread(finalizationManager.createRunner());
```

Comment:
This thread is unfortunate overhead. These are some options to avoid it:
- check the `FinalizationManager` queue before parsing starts - used by [WeakHashMap.exprungeStaleEntries()](https://github.com/openjdk/jdk/blob/47c10694c66bc131c8a5e1572340415b8daaba08/src/java.base/share/classes/java/util/WeakHashMap.java#L375)
- use [Cleaner](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/ref/Cleaner.html)
- leave the cleanup up to the user - e.g. continue to provide instances and `close()` operations

Own `Thread` adds overhead. I am not sure if there is a private thread behind `Cleaner` (then it'd be the same), or if the thread is shared. In such case it would be more effective. `exprungingStaledParser` might be good, but may leave some of the latest parser(s) pending for cleanup. We do get a [shutdown callback in the interpreter](https://github.com/enso-org/enso/blob/2fbd0aa92b7cb0c557dc37480ddb267ce8ad1fb9/engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoContext.java#L290) - e.g. we can explicitly request cleanup of `EnsoContext.getCompiler()` parser workers...


#### Option 1: Cleaner

Shared `Cleaner` avoids the overhead of own `Thread` and otherwise lets GC do its work.

#### Option 2: Exprunging stale & Close

Having `exprungeStaleParsers()` check and a way to trigger it (or request shutdown of all existing) when `EnsoContext.shutdown()` would do the trick too without need for any special `Thread` or decisions of GC.

Reply:
Can we name it? I often look into thread dump and hunting down unnamed ones is a real PITA.